pios: spi, irq,... lots of cleanup

mlyle opened the pull request:
- SPI: eliminate disused DMA callback arg
- SPI: remove DMA from F3 and F4.  wasn't used on F3, was no advantage as
  written on F4.
- SPI: check much more state / wait much more on F3... because what we were
  doing just wasn't healthy.
- SPI: completely unify F1, F3 and F4 implementations
- IRQ: unify
- debug: unify
- pwm/ppm: unify
- wdg: unify
- remove some conditional compilation things that are needed in any case
- factor STM32 specific (but not F1/F3/F4 specific) code into separate dir
- Change how makefiles find code to automatically take the "most specific" implementation

Fixes #1001
Fixes #983

mlyle committed:
pios_spi: refactor, cleanup

* eliminate disused DMA callback arg
* remove DMA from F3 and F4.  wasn't used on F3, was no advantage as
written on F4.
* make F3 and F4 implementations much closer to unified
* check much more state / wait much more on F3... because what we were
doing just wasn't healthy.

mlyle commented:
Completely untested.

mlyle committed:
pios: factor pios_spi completely common

mlyle committed:
flight: pios_spi: remove use_crc and the dma config

mlyle committed:
pios: remove PIOS_INCLUDE_IRQ

This is pretty fundamental and required for just about any other
subsystem to work.

mlyle committed:
pios: remove PIOS_INCLUDE_DELAY

mlyle committed:
pios: factor pios_irq.c to common

mluessi commented:
What's the motivation for removing DMA? For things like sending data over SPI to OpenLager, having DMA support will be useful.

mlyle commented:
As it's written, DMA is never used (transfers don't hit the threshold, no callers use the callback mechanism-- so even if they hit the threshold the CPU would busy-wait polling registers for it to be complete) and there appear to be obvious bugs in it.  There's no implementation on F3, it looks like there's multiple bugs in it on F4.

It prevents the unification of the different pios_spi implementations, uses memory and code space, etc.

Also, it uses up tons of DMA channels that we could instead use to support DMA on pios_com_usart which would be a more immediate benefit to openlager on more targets.

I'm not opposed to putting a DMA implementation back, if we can find a use case that it actually improves and we can build it correctly in a unified fashion across the stm targets.

mlyle committed:
pios: factor debug, ppm, pwm common

mluessi commented:
Makes sense, it may be good to clean it up and then add things like DMA when we actually use it.

In general, I'm not sure if "if we can build it correctly in a unified fashion across the stm targets" should be a requirement, i.e., we don't want to be forced to only use features that are supported on F1, F3, and F4.

mlyle commented:
When the code is 90%+ common, it should be in one file and have the different bits conditionalized.

mlyle committed:
pios_wdg: factor common

mlyle committed:
pios: unify pios_reset

mlyle committed:
flight: rearrange code further

Move STM32-generic things from Common to STM32.

Teach the make environment to do intelligent searching of subpaths.

Teach the make environment some about architectural details.

Simplify makefiles as a result.

mlyle committed:
flight: targets: remove PYMODULES stuff

mlyle commented:
Rebased from conflict-- pios_mpu changes

mlyle commented:
Benchtested Sparky1, Revolution, and CC3D (one of each generation).  Happy.

tracernz commented on flight/PiOS/STM32/pios_irq.c:
```diff
@@ -49,6 +47,7 @@ static uint32_t prev_primask;
 */
 int32_t PIOS_IRQ_Disable(void)
 {
+	/* XXX TODO: This is racy */
```
Nothing stopping this func. being re-ordered?

mlyle replied:
It requires an atomic / analysis (because there could be an intervening task switch).  I just spotted it while I was here, but thinking of the correct fix takes some effort.

tracernz commented:
Looks reasonable on first pass. Not really a fan of vpath though.

mlyle commented:
I had it explicitly searching the paths as part of the template before, but it was wordier and probably not as quick as letting make do the work.

Main question is whether the concept is OK-- having "general" implementations and then allowing them to be overridden by arch-specific ones without any makefile surgery.  Hopefully this makes a lot more lines of makefile common so more work can be done to factor them out.

tracernz commented:
The reason vpath could be an issue is because Re1 will require a different pios_ws2812.c than other F4 boards.

mlyle commented:
If it's completely independent, that's fine--  vpath is defined to operate in the order the paths are specified. 

> If several vpath patterns match the dependency file's name, then make processes each matching vpath directive one by one, searching all the directories mentioned in each directive. make handles multiple vpath directives in the order in which they appear in the makefile; multiple directives with the same pattern are independent of each other.

So,

```
vpath % $(BOARD_ROOT_DIR)/$(BUILD_TYPE)

vpath % $(foreach d,$(ARCH_TYPES) Common,$(PIOS)/$(d))
```

will first look in the fw directory, then PiOS/STM32F4xx, then PiOS/STM32, then PiOS/Common.   So there could be two pios_ws2812.c's, one in brainre1/fw and then one in PiOS/STM32F4xx.  (perhaps in the future we should have brainre1/common, but that's for the future IMO).

In fact, if I had a concern, it'd be that the ease of this will make copying code and making a couple changes for a single target more tempting, because it's easier.

tracernz commented:
Need to update used DMA channels table on pios_board.h for all affected targets.

mlyle committed:
pipxtreme: remove TESTAPP stuff (unused)

mlyle committed:
targets: fix up DMA usage tables

mlyle commented:
OK.  removed the SPI stuff from the DMA tables.  It was missing/lies on F1 anyways.

Brain's DMA table information is a lie, and I think RE1's may be (is QSPI DMA?)  But IMO that can wait as it doesn't relate to this changeset.

mlyle commented:
Tried to cram some other srtuff in here but pulled it out.

mluessi commented:
> Brain's DMA table information is a lie, and I think RE1's may be (is QSPI DMA?) But IMO that can wait as it doesn't relate to this changeset.

They both use DMA for the OSD SPI / QSPI, but the table is likely wrong/outdated. I can fix this in a separate PR.

mlyle committed:
pios_led: factor common

The code was nearly identical, except:

* STM32F1 had code to enable GPIO clocks.  Not necessary-- they were
already enabled.  Disabled.
* Remap is simpler on F1, so we choose how to do that based on arch.

tracernz commented:
CC3D doesn't boot firmware or upgrader with this. Quanton and Sparky1 booted okay.

mlyle commented:
Ick.  I will run that down.  I tried CC3D not too long ago, so it must have been a regression in the most recent commits.

tracernz commented:
Turns out the breakage is actually #1058.

tracernz commented:
This moves a long way in the right direction, nice work. 👍
